src/workflow/mod.rs

```diff
@@ -7,15 +7,31 @@ pub(crate) use concurrency_manager::WorkflowConcurrencyManager;
 pub(crate) use driven_workflow::{ActivationListener, DrivenWorkflow, WorkflowFetcher};
 
 use crate::{
-    machines::{ProtoCommand, WFCommand, WorkflowMachines},
+    machines::{ProtoCommand, WFCommand, WFMachinesError, WorkflowMachines},
     protos::{
         coresdk::workflow_activation::WfActivation,
-        temporal::api::{history::v1::History, workflowservice::v1::PollWorkflowTaskQueueResponse},
+        temporal::api::{common::v1::WorkflowExecution, history::v1::History},
     },
-    protosext::HistoryInfo,
-    CoreError, Result,
+    protosext::{HistoryInfo, HistoryInfoError},
 };
-use std::sync::mpsc::Sender;
+use std::sync::mpsc::{SendError, Sender};
+
+type Result<T, E = WorkflowError> = std::result::Result<T, E>;
+
+/// Errors relating to workflow management and machine logic
+#[derive(thiserror::Error, Debug, displaydoc::Display)]
+#[allow(clippy::large_enum_variant)]
+// NOTE: Docstrings take the place of #[error("xxxx")] here b/c of displaydoc
+pub enum WorkflowError {
+    /// The machine with `run_id` was not found in memory
+    MissingMachine { run_id: String },
+    /// Underlying error in state machines: {0:?}
+    UnderlyingMachinesError(#[from] WFMachinesError),
+    /// There was an error in the history associated with the workflow: {0:?}
+    HistoryError(#[from] HistoryInfoError),
+    /// Error buffering commands
+    CantSendCommands(#[from] SendError<Vec<WFCommand>>),
```

Comment:
Should we name it more imperatively, putting subject (Command) first? (e.g. CommandDispatchError)

Reply:
Yep, makes sense.